src/scheduler.ml

```diff
@@ -371,6 +432,66 @@ end = struct
   let init () = Lazy.force init
 end
 
+module Signal_watcher : sig
+  val init : unit -> unit
+end = struct
+
+  let signos = List.map Signal.all ~f:Signal.to_int
+
+  let warning = {|
+
+*************************************************************
+* Press Control+C again quicly to perform an emergency exit *
+*************************************************************
+
+|}
+
+  external sys_exit : int -> _ = "caml_sys_exit"
+
+  let signal_waiter () =
+    if Sys.win32 then begin
+      let r, w = Unix.pipe () in
+      let buf = Bytes.create 1 in
+      Sys.set_signal Sys.sigint
+        (Signal_handle (fun _ -> assert (Unix.write w buf 0 1 = 1)));
+      fun () ->
+        assert (Unix.read r buf 0 1 = 1);
+        Signal.Int
+    end else begin
+      ignore (Unix.sigprocmask SIG_BLOCK signos : int list);
+      fun () ->
+        Thread.wait_signal signos
+        |> Signal.of_int
+        |> Option.value_exn
+    end
+
+  let run wait_signal =
+    let last_exit_signals = Queue.create () in
+    while true do
+      let signal = wait_signal () in
+      Event.send_signal signal;
+      match signal with
+      | Int | Quit | Term ->
+        let now = Unix.gettimeofday () in
+        Queue.push now last_exit_signals;
+        (* Discard old signals *)
+        while Queue.length last_exit_signals >= 0 &&
+              now -. Queue.peek last_exit_signals > 1.
+        do
+          ignore (Queue.pop last_exit_signals : float)
+        done;
+        let n = Queue.length last_exit_signals in
+        if n = 2 then prerr_endline warning;
+        if n = 3 then sys_exit 1
+    done
+
+  let init = lazy (
+    let wait_signal = signal_waiter () in
+    ignore (Thread.create run wait_signal : Thread.t))
```

Comment:
Could you comment on as to why we can't just call `wait_signal` in the `run` thread.

Reply:
It's because signals must be blocked before starting any thread, so that only the thread calling `Thread.wait_signal` may receive signals

Reply:
I reworked the code to make this more explicit